### PR TITLE
Feat/cvi sde

### PR DESCRIPTION
**Pull request by prakharverma**
1. Adding CVI-SDE model.
2. Minor shape error in SDE diffusion class solved.
3. Related utils added in SDE utils class.
4. Other Minor things.

### CODE REVIEW

File: markovflow/models/variational_cvi_sde.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+from markovflow.sde.sde_utils import linearize_sde
+from markovflow.emission_model import EmissionModel
+from markovflow.kalman_filter import KalmanFilterWithSparseSites
+from markovflow.sde.sde_utils import KL_sde
+from markovflow.models.variational_cvi import gradient_transformation_mean_var_to_expectation
+
+
+class CVISDESparseSites(MarkovFlowModel):
+    """
+    Provides a site-based parameterization to the variational posterior of a dynamical system.
```
Comment (vincentadam87) [inline markovflow/models/variational_cvi_sde.py]:
```suggestion
    Provides a site-based parameterization to the variational posterior 
    over the state trajectory of a dynamical system.
```

File: tests/unit/test_sde.py [hunk 2]
```diff
@@ -141,3 +140,48 @@ def test_deterministic_euler_maruyama_value(setup):
...
 
     np.testing.assert_allclose(simulated_values, expected_values, atol=1e-5)
+
+
+def test_KL_sde(setup):
+    """
+    Test the KL_sde function which calculates the KL divergence between two SDEs with same diffusion using
+    Girsanov theorem.
+
+    We assert with decimal place 2 as KL_SDE is an approximation. With finer grid the tolerance can be made much tighter but it results in high running time.
```
Comment (vincentadam87) [inline tests/unit/test_sde.py]:
```suggestion
    We assert with decimal place 2 as KL_SDE is an approximation. 
    With finer grid the tolerance can be made much tighter but it results in high running time.
```

File: markovflow/models/variational_cvi_sde.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+        self.obs_sites_indices = tf.where(tf.equal(self.grid[..., None], self._observations_time_points))[:, 0][
+            ..., None]
+        self.dt = self.grid[1] - self.grid[0]
+
+    def _initialize_mean_statistic(self):
+        """
+        Initialize the prior on the initial state and the posterior statistics.
+        """
+        self.initial_mean = tf.zeros((1, self.state_dim), dtype=self._observations.dtype)
+        self.initial_chol_cov = tf.linalg.cholesky(tf.reshape(self.prior_sde.q, (1, self.state_dim, self.state_dim)))
```
Comment (vincentadam87) [inline markovflow/models/variational_cvi_sde.py]: what is the initial_chol_cov set to the prior diffusion?
It would be better to have a setter,
and potentially pass the arguments to the model init.

File: markovflow/models/variational_cvi_sde.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+            * Observations with shape ``batch_shape + [num_data, observation_dim]``
+
+        :param likelihood: A likelihood with shape ``batch_shape + [num_inducing]``.
+        :param learning_rate: The learning rate of the algorithm.
+        :param test_data: Test data used to calculate NLPD if not None.
+        """
+        super().__init__()
+
+        self._likelihood = likelihood
+        self.grid = grid
```
Comment (vincentadam87) [inline markovflow/models/variational_cvi_sde.py]: time_grid? would be a better name

File: markovflow/models/variational_cvi_sde.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+        grads = g.gradient(local_obj, [Fmu, Fvar])
+        # turn into gradient wrt μ, σ² + μ²
+        grads = gradient_transformation_mean_var_to_expectation((Fmu, Fvar), grads)
+
+        return local_obj, grads
+
+    def cross_term(self, dist_q: StateSpaceModel = None) -> TensorType:
+        """
+        Calculate
+                    E_{q(x)[(f_q - f_L)\Sigma^-1(f_L - f_p)] .
```
Comment (vincentadam87) [inline markovflow/models/variational_cvi_sde.py]:
```suggestion
                    E_{q(x)}[(f_q - f_L)\Sigma^-1(f_L - f_p)] .
```

File: markovflow/models/variational_cvi_sde.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+            local_obj = tf.reduce_sum(
+                input_tensor=self.local_objective(Fmu, Fvar, self._observations)
+            )
+        grads = g.gradient(local_obj, [Fmu, Fvar])
+        # turn into gradient wrt μ, σ² + μ²
+        grads = gradient_transformation_mean_var_to_expectation((Fmu, Fvar), grads)
+
+        return local_obj, grads
+
+    def cross_term(self, dist_q: StateSpaceModel = None) -> TensorType:
```
Comment (vincentadam87) [inline markovflow/models/variational_cvi_sde.py]: maybe it would be good to call this kl_cross_term
and to remind the kl expression in the docstrings and how it split.
I m thinking about the other people who might read this in the future

File: markovflow/models/variational_cvi_sde.py [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+        diag_quad = NDiagGHQuadrature(self.prior_sde.state_dim, 20)
+        q_mean = tf.squeeze(self.fx_mus, axis=0)[:-1]
+        q_covar = tf.squeeze(self.fx_covs, axis=0)[:-1]
+
+        val = diag_quad(func, q_mean, tf.squeeze(q_covar, axis=-1))
+        val = tf.reduce_sum(val) * self.dt
+
+        return val
+
+    def loss_lin(self, dist_p: StateSpaceModel = None, m: TensorType = None, S: TensorType = None):
```
Comment (vincentadam87) [inline markovflow/models/variational_cvi_sde.py]: same, kl_linearization_loss ? 
and explain the name: the loss minimized for posterior linearization

File: markovflow/models/variational_cvi_sde.py [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+        A = tf.squeeze(dist_p.state_transitions, axis=0)
+        b = tf.squeeze(dist_p.state_offsets, axis=0)
+        A = (A - tf.eye(self.state_dim, dtype=A.dtype)) / self.dt
+        b = b / self.dt
+
+        lin_loss = KL_sde(self.prior_sde, A, b, m, S, dt=self.dt)
+        return lin_loss
+
+    def update_sites(self):
+        """Update the data-sites"""
```
Comment (vincentadam87) [inline markovflow/models/variational_cvi_sde.py]:
```suggestion
        """Update of the data-sites following CVI"""
```

File: markovflow/sde/sde_utils.py [hunk 9]
```diff
@@ -133,3 +132,65 @@ def linearize_sde(
         state_offsets=state_offsets,
         chol_process_covariances=chol_process_covariances,
     )
+
+
+def KL_sde(sde_p: SDE, A_q: TensorType, b_q: TensorType, m: TensorType, S: TensorType, dt: float,
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py]: Thinking about names again
KL_mgp_to_sde?
in practice it is KL(q|p) with q a markovian gp
KL_sde is a bit too generic

Comment (vincentadam87) [reply markovflow/sde/sde_utils.py]: maybe the arguments could be m_q, S_q?
Is there a reasons why you don't pass a ssm?
Then you could extract A_q, b_q and compute m, S from it.

File: markovflow/models/variational_cvi_sde.py [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+
+        # update
+        new_data_nat1 = (1 - self.sites_lr) * self.sites.nat1 + self.sites_lr * grads[0]
+        new_data_nat2 = (1 - self.sites_lr) * self.sites.nat2 + self.sites_lr * grads[1][..., None]
+
+        self.sites.nat1.assign(new_data_nat1)
+        self.sites.nat2.assign(new_data_nat2)
+
+    def variational_expectation(self, fx_mus: TensorType = None, fx_covs: TensorType = None) -> TensorType:
+        """Likelihood variational expectation"""
```
Comment (vincentadam87) [inline markovflow/models/variational_cvi_sde.py]:
```suggestion
        """Expected log-likelihood under the current variational posterior"""
```

File: tests/integration/models/test_variational_cvi_sde.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    gpr_nat1 = gpr.observations / s2
+    gpr_nat2 = -0.5 / s2 * np.ones_like(cvi_sde_nat2)
+
+    np.testing.assert_allclose(cvi_sde_nat1, gpr_nat1)
+    np.testing.assert_allclose(cvi_sde_nat2, gpr_nat2)
+
+
+def test_elbo_terms(with_tf_random_seed, cvi_sde_gpr_optim_setup):
+    """
+    Test elbo terms
```
Comment (vincentadam87) [inline tests/integration/models/test_variational_cvi_sde.py]:
```suggestion
    For a linear SDE prior (i.e. a Gaussian Process [GP]), test that the ELBO computed
    using girsanov theorem or the usual formula for GPs match. 
```

**Review by vincentadam87 — COMMENTED** [review]
first batch of comments

File: markovflow/models/variational_cvi_sde.py (hunk 3 above)
Comment (prakharverma) [reply markovflow/models/variational_cvi_sde.py]: So, this is where I was a bit confused. In this function, we initialize the prior on the initial state by `self.initial_mean, self.initial_chol_cov` and the initial posterior `self.fx_mus, self.fx_covs` that is used to linearize the prior at the start. 

How should they be initialized? Like passed in the init with the default value being?

File: markovflow/models/variational_cvi_sde.py (hunk 6 above)
Comment (prakharverma) [reply markovflow/models/variational_cvi_sde.py]: Makes sense! I will rename the function and modify the docstring

File: markovflow/sde/sde_utils.py (hunk 9 above)
Comment (prakharverma) [reply markovflow/sde/sde_utils.py]: I agree with the name of the function being generic so will change it. :+1: 

In terms of why not passing the SSM. Basically, the A and b parameters and not always of the `q` SSM. If we think about the linear loss term as well in the CVI-SDE class where this function is used currently, we need to calculate 

`E_{q}[f_p - f_l]^2_{Sigma^{-1}}` 

so A, b are for the linear SDE and m, S are of the posterior q.

Comment (prakharverma) [reply markovflow/sde/sde_utils.py]: I see, the docstrings is a bit confusing so will modify it too.

Comment (prakharverma) [reply markovflow/sde/sde_utils.py]: The function can be used for KL[q||p] but it is more generic in a way how we use it to calculate the KL_linearization_loss term.

Comment (prakharverma) [reply markovflow/sde/sde_utils.py]: Modified the doc string and the function name for it to be more clear.

File: markovflow/models/variational_cvi_sde.py (hunk 3 above)
Comment (prakharverma) [reply markovflow/models/variational_cvi_sde.py]: I added options for it in the latest commit!

File: markovflow/models/variational_cvi_sde.py [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+        self._linearize_prior()
+
+        self.obs_sites_indices = tf.where(tf.equal(self.time_grid[..., None], self._observations_time_points))[:, 0][
+            ..., None]
+        self.dt = self.time_grid[1] - self.time_grid[0]
+
+    def _initialize_initial_state_prior(self, prior_initial_state: Gaussian):
+        """Initialize the prior on the initial state."""
+        if prior_initial_state is None:
+            prior_initial_state = Gaussian(mu=tf.zeros((1, self.state_dim), dtype=self._observations.dtype),
```
Comment (vincentadam87) [inline markovflow/models/variational_cvi_sde.py]: Any reason not to store the gpflow Gaussian object? rather than the two moments?
Here you pass or build a gaussian to later extract and store the mean and covariance

File: markovflow/sde/sde_utils.py (hunk 9 above)
Comment (vincentadam87) [reply markovflow/sde/sde_utils.py]: `squared_drift_difference_along_Gaussian_path` ? 
How about you pass `m,S` into a Gaussian object (broadcasting)
And you create a Similar wrapper class for a `LinearDrift` where you pass the A, b (broadcasting)

Comment (vincentadam87) [reply markovflow/sde/sde_utils.py]: This is really just about not having too many variables in the arguments list

File: markovflow/sde/sde_utils.py [hunk 13]
```diff
@@ -133,3 +133,65 @@ def linearize_sde(
...
+    """
+    Drift difference between the two SDEs where the first SDE can be any arbitrary SDE and the second SDE is linear,
+    with the drift defined as f(x(t)) = A(t) x(t) + b(t), along the path defined by the Gaussian with mean m and
+    (covariance) S.
+
+    Mathematically, the function calculates:
+        0.5 * E_{q}[||f_p(x(t)) - (A(t) x(t) + b(t))||^{2}_{Σ^{-1}}].
+
+    When A and b parameter is of the drift of the SDE represented by the distribution `q` whose marginals are m and S
+    then the function returns the KL[q || p].
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py]:
```suggestion
    Expected Square Drift difference between two SDEs
    * a first one denoted by p, that can be any arbitrary SDE.
    * a second which is linear, denoted by p_L, with a drift defined as f_L(x(t)) = A_L(t) x(t) + b_L(t)
    Where the expectation is over a third distribution over path
    summarized by its the mean (m) and covariance (S) for all times.

    Formally, the function calculates:
        0.5 * E_{q}[||f_p(x(t)) - (A_L(t) x(t) + b_L(t))||^{2}_{Σ^{-1}}].

    This function corresponds to the expected log density ratio:  E_q log [p_L || p].

    When A_L and b_L  are the drift parameters of `q`,
    then the function returns the KL[q || p].
```

**Review by vincentadam87 — COMMENTED** [review]
some further comments

File: markovflow/models/variational_cvi_sde.py (hunk 12 above)
Comment (prakharverma) [reply markovflow/models/variational_cvi_sde.py]: It's basically done this way because when these are used in `linearize_sde` method we will need to do casting and reshaping there so why not do it in the init. 

Now, when i think about it, linearize_sde can also take a Gaussian rather than `m` and `S`. :/

File: markovflow/sde/sde_utils.py (hunk 9 above)
Comment (prakharverma) [reply markovflow/sde/sde_utils.py]: It's done this way primarily because the method is used in cvi_sde class where we have the marginals as m and S. We will need to create a Gaussian there before calling this function.

File: markovflow/sde/sde_utils.py [hunk 14]
```diff
@@ -18,22 +18,90 @@
...
 from markovflow.sde import SDE
 from markovflow.state_space_model import StateSpaceModel
 
 
+class LinearDrift:
+    """
+    A linear drift of the form, f(x_t, t) = A_t * x_t + b_t.
+    """
+    def __init__(self, A: TensorType = None, b: TensorType = None):
+        self.A = A
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py]: docstrings? Like say it takes two tensors as arguments

File: markovflow/sde/sde_utils.py [hunk 15]
```diff
@@ -18,22 +18,90 @@
...
+        self.b = b
+
+    def from_ssm(self, ssm: StateSpaceModel, dt: float):
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py]: set_from_ssm ?
this is a setter, in the end you self.A = ..

File: markovflow/sde/sde_utils.py [hunk 16]
```diff
@@ -18,22 +18,90 @@
...
+        """
+        Convert a StateSpaceModel into a linear drift of the f(x_t, t) = A_t * x_t + b_t, where
+            A = (SSM.state_transitions - I)/dt
+            b = SSM.state_offsets / dt
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py]:
```suggestion
        """
        Approximately converts a StateSpaceModel into a linear drift of the f(x(t), t) = A(t) * x(t) + b(t),
        
        Given the linear drift f(x(t), t) = A(t) * x(t) + b(t), of a continuous time SDE
        The conditional distribution of two  states  x(t_{k+1}), x(t_{k}) is given by
        
         x(t_{k+1}) | x(t_{k}) = C_k x(t_{k}) +  b_k + n_k
         
         with C_k = exp( \int_{t_{k} < t < t_{k+1}}  A(t) dt)
         and b_k = [C_k - I] A(t)^{-1} b(t)
        
        Taking the limit of small time intervals ( dt = t_{k+1} - t_{k} -> 0)
        The first order Taylor expansion of the exponential 
        for small time intervals ( dt = t_{k+1} - t_{k} -> 0), gives the approximation:
       
        C_k ~= I + A(t_{k}) dt     and     b_k ~=  b(t_{k}) dt

        A = (SSM.state_transitions - I)/dt
        b = SSM.state_offsets / dt
```

Comment (vincentadam87) [reply markovflow/sde/sde_utils.py]: something like that, and the reverse form to_ssm.

File: markovflow/sde/sde_utils.py [hunk 17]
```diff
@@ -18,22 +18,90 @@
...
+        )
+
+        return StateSpaceModel(
+            initial_mean=initial_mean,
+            chol_initial_covariance=initial_chol_covariance,
+            state_transitions=state_transitions,
+            state_offsets=state_offsets,
+            chol_process_covariances=chol_process_covariances,
+        )
+
```
Comment (vincentadam87) [inline markovflow/sde/sde_utils.py:27]: How about a __call__(states) method where it returns the drift evaluated on the input states.

File: markovflow/sde/sde_utils.py (hunk 14 above)
Comment (vincentadam87) [reply markovflow/sde/sde_utils.py]: Maybe overkill but some shape_asserts checking that the dims match (number of time steps would be good)

We do this when we create a ssm
https://github.com/secondmind-labs/markovflow/blob/654138d0dbb67dc6c3457eadf3e2db9cd8ffd71b/markovflow/state_space_model.py#L101

Comment (prakharverma) [reply markovflow/sde/sde_utils.py]: The assert is there when these variables are used. As you mentioned, checking them multiple times would be an overkill.

File: markovflow/sde/sde_utils.py (hunk 17 above)
Comment (prakharverma) [reply markovflow/sde/sde_utils.py:27]: Actually, I had this implemented but then removed. :D
So, basically, we need `x` and corresponding `t` to pick the `A` and `b` values, right?

**Review by vincentadam87 — APPROVED** [review]
Good to go
